Fixed: Dropdowns does not show hover state inside balloon toolbar

engineering-this opened the pull request:
## What is the purpose of this pull request?

Bugfix

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_contributing_code-section-tests),
[how to set the testing environment](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests) and
[how to create tests](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests-section-creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

Existing test already cover this case.

## What changes did you make?

I've presented new private method `CKEDITOR.ui.balloonToolbarView.prototype._updateStatus` which updates status of element and used in on each `richCombo` elements in balloon toolbar.

Closes #1268

f1ames commented on plugins/balloontoolbar/plugin.js:
```diff
@@ -680,6 +680,32 @@
+			 *
+			 * @since 4.9.0
+			 * @private
+			 * @param {CKEDITOR.ui.button/CKEDITOR.ui.richCombo}
```
Missing param name, should be `@param {CKEDITOR.ui.button/CKEDITOR.ui.richCombo} item`.

f1ames commented on plugins/balloontoolbar/plugin.js:
```diff
@@ -680,6 +680,32 @@
+			 * @param {CKEDITOR.ui.button/CKEDITOR.ui.richCombo}
+			 * @member CKEDITOR.ui.balloonToolbarView
+			 */
+			CKEDITOR.ui.balloonToolbarView.prototype._updateStatus = function( item ) {
```
This method is copied from `CKEDITOR.ui.richCombo`. I see two problems with it:

1. It is exact copy and if something changes in a source method (well, not probable, but may happen), it may stop working.
2. `Item` param type seems to be `CKEDITOR.ui.button` or `CKEDITOR.ui.richCombo`, however this method is copied from `CKEDITOR.ui.richCombo` so it will not work for buttons. As this method is only used for richCombo maybe it should only accept only `CKEDITOR.ui.richCombo` type?

As for **1.** my suggestion is to make `updateStatus` method in `CKEDITOR.ui.richCombo` public (or add some new method which will internally call `updateStatus`) and then you could get rid of this method entirely which will simplify the code.

f1ames commented on tests/plugins/balloontoolbar/combostate.js:
```diff
@@ -0,0 +1,30 @@
+
+			panel.addItems( {
+					Styles: editor.ui.create( 'Styles' )
+				} );
```
This two line should have one level less of indentation:

```
panel.addItems( {
	Styles: editor.ui.create( 'Styles' )
} );
```

f1ames commented on tests/plugins/balloontoolbar/manual/dropdownhover.html:
```diff
@@ -0,0 +1,44 @@
+		bender.ignore();
+	}
+
+	CKEDITOR.disableAutoInline = true;
```
There is no inline editor (or anything which would become one), so this line looks redundant.

f1ames commented on tests/plugins/balloontoolbar/manual/dropdownhover.html:
```diff
@@ -0,0 +1,44 @@
+			Styles: evt.editor.ui.create( 'Styles' )
+		} );
+
+			strong.on( 'click', function() {
```
Also here is something wrong with indentation.

f1ames commented on tests/plugins/balloontoolbar/combostate.js:
```diff
@@ -0,0 +1,30 @@
+	bender.editor = {};
+
+	var tests = {
+		'test stylescombo state': function() {
```
This tests should be also ignored for `IE < 9` IMHO.

engineering-this commented:
> If this is the case, unit tests should be updated as it passes and doesn't cover this case (or maybe another test could be added covering it).
@f1ames 
Little tweak to test fixed this case.

Comandeer commented:
Change base to `master` and rebased.

Comandeer commented on tests/plugins/balloontoolbar/manual/dropdownhover.md:
```diff
@@ -0,0 +1,9 @@
+@bender-ui: collapsed
+@bender-tags: 4.9.1, bug, 1268, balloontoolbar
```
Please update version tag.

Comandeer commented on plugins/balloontoolbar/plugin.js:
```diff
@@ -760,6 +760,12 @@
 					element.setAttribute( 'draggable', 'false' );
 					this.registerFocusable( element );
 				}, this );
+				// We need to initially set status of richCombo items #1268.
```
Can't it be automated by e.g. forcing selection change via [`editor.forceNextSelectionCheck`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editor.html#method-forceNextSelectionCheck)?

engineering-this replied:
Doesn't work for inline editors.

engineering-this commented:
Changed base to `next` and rebased.
Fixed broken styles combo.

Comandeer commented on plugins/richcombo/plugin.js:
```diff
@@ -105,6 +105,28 @@ CKEDITOR.plugins.add( 'richcombo', {
+			 */
+			updateState: function( editor ) {
+				// Don't change state while richcombo is active (https://dev.ckeditor.com/ticket/11793).
+				if ( this.getState() == CKEDITOR.TRISTATE_ON )
```
Please update the whole function to be compliant with our code style.

Comandeer commented on tests/plugins/balloontoolbar/manual/dropdownhover.md:
```diff
@@ -0,0 +1,12 @@
+
+## For each editor
+
+1. Click on bold text two times.
```
There should be also steps for the bug that I reported during the review, so the one with refreshing state after applying styles.

engineering-this commented on plugins/balloontoolbar/plugin.js:
```diff
@@ -760,6 +760,12 @@
 				}, this );
+				// We need to initially set status of richCombo items (#1268).
+				CKEDITOR.tools.array.forEach( keys, function( itemKey ) {
+					if ( CKEDITOR.ui.richCombo && items[ itemKey ] instanceof CKEDITOR.ui.richCombo ) {
```
Actually we are iterating to render items, and adding some extra logic to rich combo. There are still possible other items like buttons, that can be rendered.

engineering-this commented:
Rebased onto latest `master`.

Comandeer commented on tests/plugins/richcombo/richcombo.js:
```diff
@@ -56,5 +56,56 @@ bender.test( {
+	'test updatestate read only editor': testUpdateState( { readOnly: true } ),
+
+	// (#1268)
+	'test updatestate source mode': testUpdateState( { mode: 'source' } ),
```
This test does not check if everything works fine in `source` mode. In reality it checks how richcombo behaves in readonly editor. To really check how it behaves in different mode, [`editor#setMode method](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editor.html#method-setMode) should be used.

Comandeer commented on tests/plugins/richcombo/richcombo.js:
```diff
@@ -56,5 +56,56 @@ bender.test( {
+		var editor = this.editor,
+			combo = editor.ui.get( 'custom_combo' ),
+			spy = sinon.spy( combo, 'setState' ),
+			expected = [ 1, CKEDITOR.TRISTATE_OFF ],
```
Wouldn't it be more readable as object with `spyCallCount` and `state` keys?

Comandeer commented on tests/plugins/richcombo/richcombo.js:
```diff
@@ -56,5 +56,56 @@ bender.test( {
+
+		if ( options.mode && options.mode !== editor.mode ) {
+			originalMode = editor.mode;
+			editor.mode = options.mode;
```
It doesn't change the mod of the editor, it forces editor to think that mode is changed, but in reality it stays the same.

Comandeer commented on tests/plugins/richcombo/richcombo.js:
```diff
@@ -56,5 +56,56 @@ bender.test( {
+
+		if ( options.readOnly || originalMode ) {
+			editor.setReadOnly( true );
+			expected[ 1 ] = CKEDITOR.TRISTATE_DISABLED;
```
What if someone tries to change from `source` to `wysiwyg` mode?

Comandeer commented on tests/plugins/balloontoolbar/combostate.js:
```diff
@@ -0,0 +1,46 @@
+	var tests = {
+		init: function() {
+			if ( CKEDITOR.env.ie && CKEDITOR.env.version < 9 || bender.tools.env.mobile ) {
+				bender.ignore();
```
It does not stop the execution of tests – they still run in IE8, ignoring the big alert that tests are ignored in current browser. You should use `assert.ignore` instead.

Comandeer commented on plugins/richcombo/plugin.js:
```diff
@@ -107,7 +107,33 @@ CKEDITOR.plugins.add( 'richcombo', {
+			 * Updates state of the combo.
+			 *
+			 * @since 4.11.0
+			 * @param {CKEDITOR.editor} editor The editor instance which this rich combo is to be used by.
```
Maybe something like

> The editor instance, in which this rich combo is located.

or

> The editor instance, which owns this rich combo.

engineering-this commented:
I have reworked unit test.
1. Now each of TC will first assert initial state of combo, so I removed TC that previously checked it.
2. Each TC will set given state/readOnly/etc. and assert combo state, eg. `wysiwyg` -> `source`.
3. After that Each TC will revert changes from step 2. and assert again. eg. `source` -> `wysiwyg`.

I've removed spy, as we are only important thing to check is combo having proper state.

Comandeer commented on tests/plugins/balloontoolbar/combostate.js:
```diff
@@ -0,0 +1,45 @@
+			if ( CKEDITOR.env.ie || CKEDITOR.env.edge ) {
+				comboButton.$.onmouseup();
+			} else {
+				comboButton.$.click();
```
This whole code can be simplified to the ternary operator:

```javascript
comboButton.$[ CKEDITOR.env.ie ? 'onmouseup' : 'onclick' ]();
```

Comandeer commented on plugins/balloontoolbar/plugin.js:
```diff
@@ -767,6 +767,14 @@
+						if ( items[ itemKey ] instanceof CKEDITOR.ui.richCombo ) {
+							items[ itemKey ].updateState( editor );
+						}
+					}, this );
```
Is passing `this` necessary?

Comandeer commented on tests/plugins/balloontoolbar/manual/dropdownhover.md:
```diff
@@ -0,0 +1,12 @@
+@bender-ui: collapsed
+@bender-tags: 4.11.2, bug, 1268, balloontoolbar
```
Please update version tag.

engineering-this commented on tests/plugins/balloontoolbar/combostate.js:
```diff
@@ -0,0 +1,45 @@
+			}
+
+			var editor = this.editor,
+				panel,
```
`comboButton` variable can be initialized after toolbar is attached.

engineering-this commented:
Applied suggested changes, reworked `testComboStates` it includes also async test steps which is needed for one case.
